runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -6937,16 +6937,71 @@ J9::Z::TreeEvaluator::VMifInstanceOfEvaluator(TR::Node * node, TR::CodeGenerator
    return NULL;
    }
 
+/**
+    * Generates a quick runtime test for valueType node and in case if node is of valueType, generates a branch to helper call
+    * 
+    * @param node monent or monexit node
+    * @param mergeLabel TR::LabelSymbol* to pointing to mergePoint in mainline
+    * @param callLabel TR::LabelSymbol* pointing to the helper Call Label
+    * @param cg CodeGenerator Object Ptr
+    * @return TR::Register* If Given node required a runtime test to check for valueType, it will return the register containing J9Class of given object
+    */
+static TR::Register*
+generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::LabelSymbol* mergeLabel, TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
+   {
+   TR::Register *objReg = cg->evaluate(node->getFirstChild());
+   TR::Register *objectClassReg = cg->allocateRegister();
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, objectClassReg, generateS390MemoryReference(objReg, TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
+   TR::Register *classFlagsRegister = cg->allocateRegister();
+   TR_J9VMBase *fej9 = static_cast<TR_J9VMBase *>(cg->comp()->fe());
+   generateRXInstruction(cg, TR::InstOpCode::L, node, classFlagsRegister, generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(fej9->getOffsetOfClassFlags()), cg));
```

Comment:
This is still not what we want. I would have hoped the use of `classFlagsRegister` would have been completely eliminated. Can we get rid of this load and just turn the `TMLL` into `TM` (immediate-memory test under mask)?

Reply:
Fixed implementation to use Memory-Immediate test instruction instead of using an additional register to hold the class flags. Also updated the `static_assert` so that if we ever change `J9ClassIsValueType` to use other bit except withing second byte of classFlags, it will throw an assert. 
Instruction looks like following. 
```
TM @(objectClassReg+(offsetOf(classFlags)+2)), 0x4
BRC COND_BRNZ, helperCall
```